clang/lib/CodeGen/CGExprScalar.cpp

```diff
@@ -2419,8 +2419,13 @@ Value *ScalarExprEmitter::VisitCastExpr(CastExpr *CE) {
     return Visit(const_cast<Expr*>(E));
 
   case CK_NoOp: {
-    return CE->changesVolatileQualification() ? EmitLoadOfLValue(CE)
-                                              : Visit(const_cast<Expr *>(E));
+    if (CE->changesVolatileQualification())
+      return EmitLoadOfLValue(CE);
+    auto V = Visit(const_cast<Expr *>(E));
+    if (CGF.CGM.getCodeGenOpts().PointerAuth.CXXMemberFunctionPointers &&
+        CE->getType()->isMemberFunctionPointerType())
+      V = CGF.CGM.getCXXABI().EmitMemberPointerConversion(CGF, CE, V);
+    return V;
```

Comment:
Why do we not need similar logic here when casting function pointer types when function pointer type diversity is enabled?  Does Sema not use `CK_NoOp` for those?  Maybe we should do the same for member pointer conversions and make sure there's a better cast kind to use.

Reply:
Sema does use `CK_NoOp` for those, but we don't need to re-sign function pointers because the presence of `noexcept` on a function type doesn't change the discriminator. The function that encodes the function type (`encodeTypeForFunctionPointerAuth`) encodes the parameter and return types, but ignores `noexcept`.

Reply:
Okay, but that's coincidental, right?  The language doesn't semantically guarantee that that conversion is a no-op, it just happens to be under normal conditions, even with function pointer type diversity enabled.

What cast kind do we use if you cast a function pointer to a completely different type?

Reply:
Depending on the kind of casts in the source code, `CStyleCastExpr` or `CXXReinterpretCastExpr` (or possibly another cast kind) is used when casting to a completely different function pointer type.

But those are explicit casts. I'm not sure there's another cast operation other than `CK_NoOp` in `clang/AST/OperationKinds.def` that we can use for that purpose.

Reply:
I mean the `CastKind`, not the expression class.  We generally treat the conversion side of `CastExpr`s uniformly in CodeGen (although there are some important exceptions).

Reply:
I don't think there's a `CastKind` we can use in `clang/AST/OperationKinds.def`. We'll have to define a new one if we are going to stop using `NoOp`.